File: paper/src/main/java/com/gitlab/aecsocket/paper/impl/PaperComponent.java
package com.gitlab.aecsocket.paper.impl;

import com.github.aecsocket.sokol.core.impl.AbstractComponent;
import com.github.aecsocket.sokol.core.stat.StatIntermediate;
import com.gitlab.aecsocket.paper.SokolPlugin;

import java.util.Map;
import java.util.Set;

public final class PaperComponent
        extends AbstractComponent<PaperComponent, PaperNodeSlot, PaperFeatureConfig> {
    private final SokolPlugin platform;

    public PaperComponent(SokolPlugin platform, String id, Set<String> tags, Map<String, PaperNodeSlot> slots, Map<String, PaperFeatureConfig> features, StatIntermediate stats) {
        super(id, tags, slots, features, stats);
        this.platform = platform;
    }

    public SokolPlugin platform() { return platform; }
}

File: paper/src/main/kotlin/com/github/aecsocket/sokol/paper/PaperComponent.kt
package com.github.aecsocket.sokol.paper

import com.github.aecsocket.sokol.core.Slot
import com.github.aecsocket.sokol.core.impl.AbstractComponent
import com.github.aecsocket.sokol.core.rule.Rule

class PaperComponent(
    id: String,
    features: Map<String, PaperFeature.Profile>,
    slots: Map<String, PaperSlot>
) : AbstractComponent<PaperComponent, PaperFeature.Profile, PaperSlot>(id, features, slots)

data class PaperSlot(
    override val key: String,
    override val tags: Set<String>,
    val required: Boolean,
    val modifiable: Boolean,
    val compatible: Rule
) : Slot
